Adjust spacing on Express Checkout buttons

asumaran opened the pull request:
Fixes https://github.com/Automattic/woocommerce-payments/issues/9169
Fixes https://github.com/Automattic/woocommerce-payments/issues/9137

#### Changes proposed in this Pull Request

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

#### Verify the spacing between express checkout buttons is consistent (`12px`) exactly as the [figma file](https://www.figma.com/design/c8VYjXqluE9vlVSE1RgJix/%5BPublic%5D-Cart-%26-Checkout-Blocks?node-id=858-15405&t=7OnWS0DaISZ1VWTb-1) on the following pages:

- <details><summary>Product page</summary><img width="820" alt="Screenshot 2024-08-14 at 2 53 45 PM" src="https://github.com/user-attachments/assets/ecd697bc-dce5-42a0-93d8-598065df30ee"></details>
- Cart block
    - <details><summary>Desktop</summary><img width="453" alt="Screenshot 2024-08-14 at 2 54 27 PM" src="https://github.com/user-attachments/assets/d02bc901-8e44-4651-a631-f3316b409c78"></details>
    - <details><summary>Mobile</summary><img width="500" src="https://github.com/user-attachments/assets/c14f5f29-2326-4dad-83d8-22dd2d1e5a76"></details>
- Checkout block
    - <details><summary>Desktop</summary><img width="805" alt="Screenshot 2024-08-14 at 2 55 50 PM" src="https://github.com/user-attachments/assets/8fa3a6de-4de3-4360-acd7-2a4d2762e84c"></details>
    - <details><summary>Mobile</summary><img width="500" src="https://github.com/user-attachments/assets/44b5dfb7-b875-43cd-bd37-108b4ed2c694"/></details>
- Cart shortcode
    - <details><summary>Desktop</summary><img width="445" alt="Screenshot 2024-08-14 at 2 58 11 PM" src="https://github.com/user-attachments/assets/baf925eb-fb87-4737-a496-bc33eef6815f"></details>
    - <details><summary>Mobile</summary><img width="500" src="https://github.com/user-attachments/assets/8f90e308-10a3-458d-acbb-6187e8b28c6a" /></details>
- Checkout shortcode
    - <details><summary>Desktop</summary><img width="812" alt="Screenshot 2024-08-14 at 3 03 38 PM" src="https://github.com/user-attachments/assets/b40d519c-c925-4b3f-965b-dde2e705c848"></details>
    - <details><summary>Mobile</summary><img width="500" src="https://github.com/user-attachments/assets/02f7de10-2b16-480a-8ac8-308bbde532ac" /></details>
- Button Sizes
    - <details><summary>Small</summary><img width="829" alt="Screenshot 2024-08-14 at 3 10 06 PM" src="https://github.com/user-attachments/assets/ba4d8356-5382-41cc-9a0c-0669c216d054"></details>
    - <details><summary>Medium</summary><img width="803" alt="Screenshot 2024-08-14 at 3 10 49 PM" src="https://github.com/user-attachments/assets/7a76c60b-e578-4a2e-8b82-51b6c395efb2"></details>
    - <details><summary>Large</summary><img width="810" alt="Screenshot 2024-08-14 at 3 14 08 PM" src="https://github.com/user-attachments/assets/fea0347b-b166-4ad2-b9de-0f551cedcecb"></details>
- Theme
    - <details><summary>Dark</summary><img width="441" alt="Screenshot 2024-08-14 at 3 24 33 PM" src="https://github.com/user-attachments/assets/da5750f1-de0f-43c0-a226-453694abdb23"></details>
    - <details><summary>Light</summary><img width="440" alt="Screenshot 2024-08-14 at 3 26 41 PM" src="https://github.com/user-attachments/assets/de3da3f8-a15a-4f85-b042-064854c558e3"></details>
    - <details><summary>Outline</summary><img width="446" alt="Screenshot 2024-08-14 at 3 27 25 PM" src="https://github.com/user-attachments/assets/04e7445f-de3d-4b8d-addc-7914ca5d3653"></details>
- Border radius
    - <details><summary>No border radius</summary><img width="444" alt="Screenshot 2024-08-14 at 3 30 53 PM" src="https://github.com/user-attachments/assets/bc72a9ef-d962-4234-9e56-e7c67f9a3487"></details>
    - <details><summary>10px border radius</summary><img width="446" alt="Screenshot 2024-08-14 at 3 33 08 PM" src="https://github.com/user-attachments/assets/00253c0e-7747-49f5-b0f1-8de545fab21a"></details>
- <details><summary>Settings - Preview</summary><img width="653" alt="Screenshot 2024-08-14 at 3 45 39 PM" src="https://github.com/user-attachments/assets/7491a060-a5e0-4050-8365-d191062219b7"></details>

#### Verify Button height is fixed (https://github.com/Automattic/woocommerce-payments/issues/9137)

- Make sure to use Jurassic Tube and Safari to be able to see Apple Pay.
- Add a product to the cart.
- Go to the block-based checkout page.
- Verify buttons are horizontally aligned 
- <img width="841" alt="Screenshot 2024-08-14 at 4 36 49 PM" src="https://github.com/user-attachments/assets/3cd6595e-3180-474b-8ef7-3ac4245d7c60">

#### Verify "OR" separator is visible when WooPay is disabled and Google/Apple Pay are enabled.

- Go to WooPayments settings and disable WooPay
- Add a product to the cart
- Go to the **shortcut-based** checkout page.
- Make sure the "OR" horizontal line is present
- <img width="822" alt="Screenshot 2024-08-14 at 4 48 54 PM" src="https://github.com/user-attachments/assets/f39593a9-3dc0-44a8-85a0-cfee77d4b55e">

-------------------

- [ ] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

asumaran committed:
No need to set a minimum height as buttons are independent now

asumaran committed:
Tweak buttons height for better alignment

asumaran committed:
Remove unnecessary import

asumaran committed:
Set 11px gap between express checkout buttons

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `9258` or branch name `as-fix-ece-spacing` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=as-fix-ece-spacing) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 40e94711eddbdee61f597f2dd229516751aee8f0
- Build time: 2024-08-16 18:23:20 UTC

_Note: the build is updated when a new commit is pushed to this PR._

asumaran committed:
Add changelog entry

asumaran committed:
Set button separator HTML ID according to ECE status

asumaran committed:
Fix failing unit test

asumaran committed:
Set 12px gap between express checkout buttons

pierorocca commented:
This is nice work @asumaran. You've provided the most comprehensive set of screenshots in recent memory. Super impressive and super helpful.

A side question, how did we get the outline on Google Pay? Are we adding it or did Stripe suddenly enable it?

asumaran commented:
>A side question, how did we get the outline on Google Pay? Are we adding it or did Stripe suddenly enable it?

@pierorocca Stripe does it inside their iframe, I've checked the Elements Appearance API and there's no way to customize it.

pierorocca commented:
![Screenshot_20240814_205231_Chrome.jpg](https://github.com/user-attachments/assets/064c98bb-8adc-435b-b020-18ed892a0db5)

Asking because before we couldn't get the outline to display and the documentation still shows as it not supported.

asumaran commented:
@pierorocca I see what you mean. Google Pay has two styles only. `Dark` (`black` on Stripe) and `Light` (`white` on Stripe). For the `Outline` theme [we are re-using](https://github.com/Automattic/woocommerce-payments/blob/106db15e504912b1561da4139d900d8fba47d284/client/express-checkout/utils/index.ts#L203-L206) the `Light` theme since it has a black border on the button.

This is how buttons look with the Light theme activated with a white background to see the difference.

<img width="507" alt="Screenshot 2024-08-14 at 11 12 42 PM" src="https://github.com/user-attachments/assets/297b3d1d-8c00-47d4-be3f-e7c1ec0b3d6e">

So to answer your question, GooglePay doesn't support the `Outline` theme, turns out Google Pay with the `Light` theme has a black border so it's being reused for the `Outline` theme.

pierorocca commented:
Ha makes total sense! lol.

ricardo commented on client/checkout/woopay/express-button/index.js:
```diff
@@ -11,7 +11,6 @@ import { getConfig } from 'utils/checkout';
 import { WoopayExpressCheckoutButton } from './woopay-express-checkout-button';
 import WCPayAPI from '../../api';
 import request from '../../utils/request';
-import '../../express-checkout-buttons.scss';
```
Just making sure: We don't need any of the ".wcpay-payment-request-wrapper" styles from `express-checkout-buttons.scss`? In case only WooPay is enabled and not Apple Pay, it won't enqueue those styles.

asumaran replied:
It's being loaded twice. I haven't investigated thoroughly, but the file is imported by another file.

ricardo replied:
Inspecting the selectors from this file, the styles are not applied if Apple Pay/Google Pay is disabled from settings.

asumaran replied:
I'll double check tomorrow.

asumaran replied:
@ricardo I don't see any issues by not loading the CSS file inside this JS file. The WooPay button is rendered alone if ApplePay/GooglePay is disabled.

ricardo replied:
Yes, it's rendered, but I wanted to make sure none of the styles around the `wcpay-payment-request-wrapper` were important for us.

Feel free to merge if you can confirm that.

asumaran replied:
I didn't observe any issue. I've checked the Product page, Cart, Checkout (both shortcode and blocks) and preview widget on the settings area.

ricardo commented on client/express-checkout/blocks/components/express-checkout-component.js:
```diff
@@ -85,6 +85,18 @@ const ExpressCheckoutComponent = ( {
 		onReady( event );
 	};
 
+	if ( buttonOptions.buttonTheme.applePay === 'black' ) {
+		if ( expressPaymentMethod === 'applePay' ) {
+			buttonOptions.buttonHeight = buttonOptions.buttonHeight + 0.4;
+		}
+	} else if (
+		expressPaymentMethod === 'googlePay' &&
+		buttonOptions.buttonTheme.googlePay === 'white'
+	) {
+		// Adjust height for Google Pay only
+		buttonOptions.buttonHeight = buttonOptions.buttonHeight - 2;
+	}
```
Nit: Minor suggestion to move this to a separate function similar to `getPaymentMethodsOverride`.

asumaran replied:
done [f6805a6](https://github.com/Automattic/woocommerce-payments/pull/9258/commits/f6805a6734249f1a274f2c26e98c9fdbf9559f92)

ricardo reviewed (approved):
LGTM and tests well. Just left 2 minor comments.

- ✅ Verify Button height is fixed (https://github.com/Automattic/woocommerce-payments/issues/9137).
- ✅ Verify "OR" separator is visible when WooPay is disabled and Google/Apple Pay are enabled.

asumaran committed:
Merge branch 'develop' into as-fix-ece-spacing

asumaran committed:
Add adjustButtonHeights to encapsulate the logic for adjusting button heights

asumaran committed:
Typo

asumaran committed:
Remove unused element instead of keep it hidden

It makes easier to query the visible elements via CSS

asumaran commented:
@ricardo do you mind taking a final look? I've addressed your comments and added a minor change to remove the button container in case it's not available.

ricardo reviewed (approved):
LGTM and tests well.

reykjalin commented on client/express-checkout/blocks/components/express-checkout-component.js:
```diff
@@ -29,6 +29,26 @@ const getPaymentMethodsOverride = ( enabledPaymentMethod ) => {
+const adjustButtonHeights = ( buttonOptions, expressPaymentMethod ) => {
+	// Apple Pay has a nearly imperceptible height difference. We increase it by 1px here.
+	if ( buttonOptions.buttonTheme.applePay === 'black' ) {
+		if ( expressPaymentMethod === 'applePay' ) {
+			buttonOptions.buttonHeight = buttonOptions.buttonHeight + 0.4;
+		}
```
@asumaran This breaks the ECE implementation when you're using the Largest size of button because `buttonHeight` is at `55` which is already the maximum allowed value to be passed to the ECE API. We need to fix this because none of the WooPayments buttons render on the max size, including the WooPay button:

<img width="378" alt="image" src="https://github.com/user-attachments/assets/95a9a6ef-aadc-47ab-9d34-04ea34090099">

pierorocca replied:
Yikes. How often would this happen?

reykjalin replied:
Anytime a merchant has ECE enabled and is using the `Large` size of the buttons. Probably not very often? Although I honestly can't say, it's hard to measure AFAIK.

asumaran replied:
I've submitted a PR fixing this. https://github.com/Automattic/woocommerce-payments/pull/9327

asumaran replied:
>Anytime a merchant has ECE enabled and is using the Large size of the buttons. 

@pierorocca, apart from that, it only happens on the block-based cart and checkout pages. It's working as usual on the product page, shortcode-based cart, and checkout pages.

alexflorisca commented:
Can I please request that these kind of changes - to the layout of ALL express payment buttons -  be done in Woo Core? It feels wrong that WooPayments is doing this. What if a user disables WooPayments? The layout would break. 

I've changed `padding-bottom` to `margin-bottom` in https://github.com/woocommerce/woocommerce/pull/50688/files to fix an issue with the layout so that clashes with this PR. If you could spin up a PR in Woo Core for these changes in the future or sync with team Rubik that would be really helpful 🙏
